rlwe/encryptor.go

```diff
@@ -0,0 +1,509 @@
+package rlwe
+
+import (
+	"github.com/ldsec/lattigo/v2/ring"
+	"github.com/ldsec/lattigo/v2/utils"
+)
+
+// Encryptor in an interface for encryptors
+//
+// encrypt with pk : ciphertext = [pk[0]*u + m + e_0, pk[1]*u + e_1]
+// encrypt with sk : ciphertext = [-a*sk + m + e, a]
+type Encryptor interface {
+	// EncryptNew encrypts the input plaintext using the stored key and returns
+	// the result on a newly created ciphertext. The encryption is done by first
+	// encrypting zero in QP, dividing by P and then adding the plaintext.
+	// The level of the output ciphertext is plaintext.Level().
+	EncryptNew(plaintext *Plaintext) *Element
+
+	// EncryptNTTNew encrypts the input plaintext using the stored key and returns
+	// the result on a newly created ciphertext. The encryption is done by first
+	// encrypting zero in QP, dividing by P and then adding the plaintext.
+	// The level of the output ciphertext is plaintext.Level().
+	// Output will be in the NTT domain.
+	EncryptNTTNew(plaintext *Plaintext) *Element
+
+	// Encrypt encrypts the input plaintext using the stored key, and returns
+	// the result on the receiver ciphertext. The encryption is done by first
+	// encrypting zero in QP, dividing by P and then adding the plaintext.
+	// The level of the output ciphertext is min(plaintext.Level(), ciphertext.Level()).
+	// Output domain will match ciphertext.Value[0].IsNTT value.
+	Encrypt(plaintext *Plaintext, ciphertext *Element)
+
+	// EncryptFastNew encrypts the input plaintext using the stored key and returns
+	// the result on a newly created ciphertext. The encryption is done by first
+	// encrypting zero in Q and then adding the plaintext.
+	// The level of the output ciphertext is plaintext.Level().
+	EncryptFastNew(plaintext *Plaintext) *Element
```

Comment:
As discussed some time ago, one way to reduce this interface would be to have two implementation of the encryptor, one for each encryption algo. Then the user chooses the type of encryption at initialization with something like: `NewEncryptor`  and `NewFastEncryptor`.

Reply:
I would rather keep a general interface and have more methods than split it into two different interfaces

Reply:
Ok, then split both and have the scheme encryptor constructor instantiate the correct one based on the provided parameters

Reply:
Done